src/main/java/com/google/sps/servlets/AddSaleCardServlet.java

```diff
@@ -0,0 +1,173 @@
+ * to this servlet.
+ */
+@WebServlet("/add-saleCard")
+public class AddSaleCardServlet extends HttpServlet {
+
+  private final static float MAX_DISTANCE = 3000f; // 3 kilometers
+  private final static float MIN_DISTANCE = 0f;
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // TODO: Get if vendor is logged in with auth API
+    boolean vendorStatus = false;
+    
+    if (vendorStatus) {
+      // -- The way we get these values might change. This is just an idea ---------------
+      // Get parameters
+      String businessName = HttpServletUtils.getParameter(request, "businessName", "");
+      boolean hasDelivery = Boolean.parseBoolean(HttpServletUtils.getParameter(request, "hasDelivery", "false"));
+      String startTime = HttpServletUtils.getParameter(request, "startTime", "");
+      String endTime = HttpServletUtils.getParameter(request, "endTime", "");
+      String description = HttpServletUtils.getParameter(request, "description", "");
+      String geoHash = HttpServletUtils.getParameter(request, "geoHash", "");
+      float radius = 0f;
+      float latitude = 0f;
+      float longitude = 0f;
+      GeoPt vendorLocation = new GeoPt(0f, 0f);
+      
+      // Check values are valid format
+      try {
+        LocalTime start = LocalTime.parse(startTime);
+        LocalTime end = LocalTime.parse(endTime);
+        radius = Float.parseFloat(HttpServletUtils.getParameter(request, "radius", "1000"));
+        latitude = Float.parseFloat(HttpServletUtils.getParameter(request, "lat", "360"));
+        longitude = Float.parseFloat(HttpServletUtils.getParameter(request, "long", "360"));
+        vendorLocation = new GeoPt(latitude, longitude);
+      } catch (DateTimeParseException e) {
+        System.out.println("Bad format to parse: " + e);
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      } catch (NumberFormatException e) {
+        System.out.println("The string is not a parsable float: " + e);
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      } catch (IllegalArgumentException e) {
+        System.out.println("Latitude and/or longitude outside legal range: " + e);
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
+
+      // Get the BlobKey of the image that the user uploaded to Blobstore
+      BlobKey imageBlobKey = getUploadedFileBlobKey(request, "imageFile");
+      String altText = HttpServletUtils.getParameter(request, "altText", "");
+
+      // Check values are not empty or outside range
+      if (businessName.isEmpty() || description.isEmpty() || geoHash.isEmpty() || altText.isEmpty() || 
+          imageBlobKey == null || radius < MIN_DISTANCE || radius > MAX_DISTANCE ) {
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
```

Comment:
Nit: Throw something in the console, save the error in a log file or just send the redirection with an error code to see the error on GCP or elsewhere.

Reply:
Done!